core/src/main/scala/sttp/client3/internal/package.scala

```diff
@@ -30,6 +34,18 @@ package object internal {
     transfer()
   }
 
+  private[client3] def withCharset(headers: Map[String, String], encoding: String): Map[String, String] = {
+    val contentTypeOpt = for {
+      charset <- Try(Charset.forName(encoding)).toOption
+      contentType <- headers.get(HeaderNames.ContentType) if !contentType.contains("charset")
```

Comment:
headers need to be compare using case-insensitive checks. There's a `Header.is` helper method, as well as a helper `Headers` wrapper, which even has a `.contentType` method on it. Maybe we can use it here.